### ISSUE TITLE
Expose terminal process ID through API

### ISSUE THREAD

**Issue by Tyriar** [Microsoft/vscode#11919]
Part of https://github.com/Microsoft/vscode/issues/11422

**Comment by Tyriar** [comment]
Extension authors want to get at the process ID of the shell so they can track what they run, at least on a low level. The problem is that due to a bug in electron (https://github.com/electron/electron/issues/38) there is [another process](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/terminal/electron-browser/terminalProcess.js) in between the renderer thread and the shell which launches the shell, meaning it's an async operation to grab the PID after the terminal is created.

Therefore, in order to maintain backwards compatibility I think this will need to be exposed as a promise:

```ts
// Returns the shell PID
getProcessID(): Thenable<number>;
```

Alternatively we could expose the in-between process synchronously, it would probably mean more work and confusion for extension authors though:

```ts
// Returns the PID of the process that launches the shell
getProcessID(): number;
```

@jrieken does the `Thenable` version look reasonable given the situation?

**Comment by jrieken** [comment]
You could just expose the tenable property? Like 

```
interface Terminal {
 // ...
 pid: Thenable<number>
}
```

**Comment by Tyriar** [comment]
If that's preferable 👍

**Comment by jrieken** [comment]
Yeah - I think that more JavaScript'ish

**Comment by Tyriar** [comment]
FYI @weinand for the debug stuff we discussed https://github.com/Microsoft/vscode/issues/9957

**Comment by Tyriar** [comment]
I ended up expanding to `processId` for clarity:

```ts
interface Terminal {
  processId: Thenable<number>
}
```

FYI @daviwil

**Comment by daviwil** [comment]
Awesome, thanks man!

**Comment by whitecolor** [comment]
@Tyriar So it is not there yet? Updated to latest insiders and terminal._id stopped to work.

**Comment by Tyriar** [comment]
@whitecolor insiders just exposed the above API `terminal.processId`. For your use case though you should use `window.onDidCloseTerminal`.

**Comment by whitecolor** [comment]
@Tyriar Ok `processId` is available now. As I've mentioned I used pid for getting number of child processes not only to determine if is opened/closed, but aslo its state (depending on number of currently running processes).

There seem to be a problemwith `processId`: it seem to expose some other `id` that was previously in `terminal._id` - number of its child processes always is zero (I use [psTree](https://github.com/indexzero/ps-tree) to get running children).

And aslo why `setProcessId` is a public method? What does it supposed to do? https://github.com/Microsoft/vscode/blob/7236f593e7e8363ac940660a105c15e4f51ea198/src/vs/workbench/api/node/extHostTerminalService.ts#L70

**Comment by Tyriar** [comment]
@whitecolor `processId` is the shell process, not the process that launches the shell:

Before: `_id` was the pid of terminalProcess which spawned bash/zsh/cmd/etc.
After: `processId` is the pid of bash/zsh/cmd/etc. that was spawned by terminalProcess

The child processes of `processId` should be the programs launched by the shell.

`setProcessId` is public on `ExtHostTerminalService` but it's not part of the API interface so you should not be touching it, [`vscode.d.ts`](https://github.com/Microsoft/vscode/blob/master/src/vs/vscode.d.ts) defines the API.

**Comment by whitecolor** [comment]
ok, but it seem that internal processes (running inside the console) is spawned not by this processesId, but by teminalProcess.

Btw `processId` is not available on the `terminal` just after  `window.createTerminal` executed, have to wait some time, possible to fix?

**Comment by jrieken** [comment]
@Tyriar Maybe we should not make `setProcessId` a public method. How about a ctor argument which is the (still unresolved) promise of the process id. That will also not require having this ugly [wait loop](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/api/node/extHostTerminalService.ts#L43)

**Comment by Tyriar** [comment]
@whitecolor that cannot change due to a bug in Electron forcing the in between process to begin with. It doesn't seem right that the processes would be children of terminalProcess.js which only spawns the shell. I looked into this in the past and it worked fine.

There's only a single spawn in terminalProcess https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/terminal/electron-browser/terminalProcess.js#L30 The processes launched by the terminal are launched by the shell that they're run in.

**Comment by Tyriar** [comment]
@jrieken :+1: #12434

**Comment by whitecolor** [comment]
@Tyriar ok, thanks,  I will look into it more carefully

**Comment by whitecolor** [comment]
@Tyriar I check all process running in windows (`wmic.exe PROCESS GET Name,ProcessId,ParentProcessId,Status`) after terminal created and I can not find there pid returned by `terminal.processId.then(...)`

**Comment by Tyriar** [comment]
Can you launch vscode, the terminal and then some program in the terminal and see what the tree looks like? I expect it to be:

- VS Code
  - ...
  - terminalProcess.js
    - cmd.exe
      - program ran in cmd.exe

**Comment by whitecolor** [comment]
> and see what the tree looks like

What tree?

**Comment by Tyriar** [comment]
@whitecolor depends on your platform, on Linux I can run `pstree`. This is the output I get, you can see that `gulp` is under `bash`:

```
        │         │         │         ├─code-insiders─┬─code-insiders───code-insiders─┬─code-insiders─┬─2*[code-insiders───{code-insiders}]
        │         │         │         │               │                               │               ├─code-insiders─┬─{WorkerPool/1113}
        │         │         │         │               │                               │               │               └─{code-insiders}
        │         │         │         │               │                               │               ├─{WorkerPool/1120}
        │         │         │         │               │                               │               └─5*[{code-insiders}]
        │         │         │         │               │                               ├─code-insiders─┬─2*[{WorkerPool/1150}]
        │         │         │         │               │                               │               ├─6*[{WorkerPool/1151}]
        │         │         │         │               │                               │               ├─{WorkerPool/9554}
        │         │         │         │               │                               │               └─5*[{code-insiders}]
        │         │         │         │               │                               ├─code-insiders─┬─{WorkerPool/1079}
        │         │         │         │               │                               │               ├─{WorkerPool/6316}
        │         │         │         │               │                               │               ├─{WorkerPool/7855}
        │         │         │         │               │                               │               └─5*[{code-insiders}]
        │         │         │         │               │                               ├─code-insiders─┬─bash───gulp─┬─4*[{V8 WorkerThread}]
        │         │         │         │               │                               │               │             ├─4*[{gulp}]
        │         │         │         │               │                               │               │             └─{node}
        │         │         │         │               │                               │               ├─2*[{WorkerPool/1149}]
        │         │         │         │               │                               │               └─2*[{code-insiders}]
        │         │         │         │               │                               ├─{Chrome_ChildIOT}
        │         │         │         │               │                               ├─3*[{CompositorTileW}]
        │         │         │         │               │                               ├─{Compositor}
        │         │         │         │               │                               ├─{HTMLParserThrea}
        │         │         │         │               │                               ├─{WorkerPool/1010}
        │         │         │         │               │                               ├─3*[{WorkerPool/1011}]
        │         │         │         │               │                               └─5*[{code-insiders}]
```

**Comment by whitecolor** [comment]
Do you see pid got from `processId` in the list?

**Comment by whitecolor** [comment]
I run `node`:
![image](https://cloud.githubusercontent.com/assets/736697/18761565/e4508ba8-811e-11e6-8bb7-b7681366d299.png)
There is pid 984

And `processId` gives 868

**Comment by whitecolor** [comment]
@Tyriar can you confirm my issue with process id?

**Comment by Tyriar** [comment]
@whitecolor it's being tested today by 2 team members that will verify the process ID.

**Comment by whitecolor** [comment]
@Tyriar Was it a bug and fixed?

**Comment by Tyriar** [comment]
@whitecolor no bug has been reported so far, not sure how many people are using it though.

**Comment by whitecolor** [comment]
> it's being tested today by 2 team members that will verify the process ID.

So it was tested and you can not confirm it?

**Comment by Tyriar** [comment]
Yes nothing was reported during the test phase and it's currently in stable.

**Comment by whitecolor** [comment]
@Tyriar Well I'm not sure how I should demonstate that something wrong here.

For me `terminal.processId` promise gives the PID value that doesn't exists in processes list at all.

![image](https://cloud.githubusercontent.com/assets/736697/19646311/48000186-9a12-11e6-849b-1cd1736f5076.png)

**Comment by Tyriar** [comment]
Created https://github.com/Microsoft/vscode/issues/14286 to track your issue
